helper/resource/state.go

```diff
@@ -180,12 +196,48 @@ func (conf *StateChangeConf) WaitForState() (interface{}, error) {
 		r := lastResult.Load().(Result)
 		return r.Result, r.Error
 	case <-time.After(conf.Timeout):
-		r := lastResult.Load().(Result)
-		return nil, &TimeoutError{
-			LastError:     r.Error,
-			LastState:     r.State,
-			Timeout:       conf.Timeout,
-			ExpectedState: conf.Target,
+		// There is special processing to handle the case where conf.Refresh is an
+		// asynchronous method that can create resources (e.g., a call to create
+		// a cloud resource). It's possible that, when we hit the timeout, the
+		// asynchronous process is actually going to succeed and provision a
+		// resource, but we'll never see it here. So, we create a mutex around the
+		// refresh attempts and then give it a grace period of 5 seconds (chosen
+		// somewhat arbitrarily) to finish processing the existing call before we
+		// give up. Note that we never actually release the mutex here in order to
+		// prevent the refresh function from getting called yet again.
+		// The length of the grace period is a tradeoff. If we don't have a grace
+		// period, we could leak resources. If we have an infinite grace period,
+		// then we might never timeout if conf.Refresh never finishes.
+		mutexLockCh := make(chan struct{})
+		go func() {
+			refreshMutex.Lock()
```

Comment:
This was somewhat pedantic, and isn't really relevant now that I see your overall goal here. 

Just to explain, technically there's no guarantee that the first goroutine is scheduled to run before this one, therefore this `refreshMutex.Lock` could succeed first short circuiting the grace period. It's not really relevant because  things are already broken if we can't run the first goroutine before the timeout, it was just a correctness issue that has bit us in other tighter scheduling cases.